Index: LightZone/lightcrafts/src/com/lightcrafts/ui/toolkit/IconFontFactory.java
===================================================================
--- LightZone.orig/lightcrafts/src/com/lightcrafts/ui/toolkit/IconFontFactory.java
+++ LightZone/lightcrafts/src/com/lightcrafts/ui/toolkit/IconFontFactory.java
@@ -5,7 +5,7 @@
 package com.lightcrafts.ui.toolkit;
 
 import jiconfont.IconCode;
-import jiconfont.icons.FontAwesome;
+import jiconfont.icons.font_awesome.FontAwesome;
 import jiconfont.icons.google_material_design_icons.GoogleMaterialDesignIcons;
 import jiconfont.swing.IconFontSwing;
 import lombok.val;
