Index: dcraw/dcraw_lz.c
===================================================================
--- dcraw.orig/dcraw_lz.c
+++ dcraw/dcraw_lz.c
@@ -9435,11 +9435,14 @@ konica_400z:
     width -= 30;
   } else if (!strcmp(make,"Sony") && raw_width == 8000) {
     width -= 32;
-    if (!strncmp(model,"DSC",3)) {
-      tiff_bps = 14;
-      load_raw = &CLASS unpacked_load_raw;
-      black = 512;
-    }
+    // This was introduced with update to dcraw 9.27,
+    // but led to broken decode for compressed files from Sony DSC-RX1RM.
+    // c.f. https://github.com/Beep6581/RawTherapee/issues/3358
+    // if (!strncmp(model,"DSC",3)) {
+    //   tiff_bps = 14;
+    //   load_raw = &CLASS unpacked_load_raw;
+    //   black = 512;
+    // }
   } else if (!strcmp(model,"DSLR-A100")) {
     if (width == 3880) {
       height--;
